internal/component/loki/source/file/internal/tail/file.go

```diff
@@ -0,0 +1,255 @@
+// from the file as they are appended, handling file events such as truncation,
+// deletion, and modification. File is safe for concurrent use.
+type File struct {
+	cfg    *Config
+	logger log.Logger
+
+	// protects file, reader, and lastOffset.
+	mu     sync.Mutex
+	file   *os.File
+	reader *bufio.Reader
+
+	lastOffset int64
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// Next reads and returns the next line from the file.
+// It blocks until a line is available, file is closed or unrecoverable error occurs.
+// If file was closed context.Canceled is returned.
+func (f *File) Next() (*Line, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.ctx.Err() != nil {
+		return nil, f.ctx.Err()
+	}
+
+read:
+	text, err := f.readLine()
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			if err := f.wait(text != ""); err != nil {
+				return nil, err
+			}
+			goto read
+		}
+		return nil, err
+	}
+
+	offset, err := f.offset()
+	if err != nil {
+		return nil, err
+	}
+
+	f.lastOffset = offset
+
+	return &Line{
+		Text:   text,
+		Offset: offset,
+		Time:   time.Now(),
+	}, nil
+}
+
+// Size returns the current size of the file in bytes.
+// It is safe to call concurrently with other File methods.
+func (f *File) Size() (int64, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	fi, err := f.file.Stat()
+	if err != nil {
+		return 0, err
+	}
+	return fi.Size(), nil
+}
+
+// Stop closes the file and cancels any ongoing wait operations.
+// After Stop is called, Next() will return errors for any subsequent calls.
+// It is safe to call Stop multiple times.
+func (f *File) Stop() error {
+	f.cancel()
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	return f.file.Close()
+}
+
+// wait blocks until a file event is detected (modification, truncation, or deletion).
+// Returns an error if the context is canceled or an unrecoverable error occurs.
+func (f *File) wait(partial bool) error {
+	offset, err := f.offset()
+	if err != nil {
+		return err
+	}
+
+	event, err := blockUntilEvent(f.ctx, f.file, offset, f.cfg)
+	switch event {
+	case eventModified:
+		if partial {
+			// We need to reset to last successful offset because we consumed a partial line.
+			f.file.Seek(f.lastOffset, io.SeekStart)
+			f.reader.Reset(f.file)
+		}
+		return nil
+	case eventTruncated:
+		// We need to reopen the file when it was truncated.
+		return f.reopen(true)
+	case eventDeleted:
+		// In polling mode we could miss events when a file is deleted, so before we give up
+		// we try to reopen the file.
+		return f.reopen(false)
+	default:
+		return err
+	}
+}
+
+// readLine reads a single line from the file, including the newline character.
+// The newline and any trailing carriage return (for Windows line endings) are stripped.
+func (f *File) readLine() (string, error) {
+	line, err := f.reader.ReadString('\n')
+	if err != nil {
+		return line, err
+	}
+
+	line = strings.TrimRight(line, "\n")
+	// Trim Windows line endings
+	line = strings.TrimSuffix(line, "\r")
```

Comment:
Why not just TrimRight "\r\n"? TrimRight will remove all of the trailing characters that are in the set, so it'll remove anything, and the current behavior will only trim a single newline on Windows but all newlines on non-Windows.